accepted/0000-automatic-migration-salt-bundle.md

```diff
@@ -0,0 +1,62 @@
+- Feature Name: Automatic migration from Salt 3000 to Salt Bundle
+- Start Date: 2023-03-22
+
+# Summary
+[summary]: #summary
+
+Force an automatic migration for Salt clients from Salt 3000 to Salt Bundle.
+
+
+# Motivation
+[motivation]: #motivation
+
+
+Salt 3000 is EOL as of Aug 31, 2022. It is advisable for Salt clients using Salt 3000 to migrate to Salt Bundle. Right now this migration can be done by applying the already existent `util.mgr_switch_to_venv_minion` state from the CLI. The puprose of this RFC is to automatically perform this migration for Salt 3000 clients to Salt Bundle so they can keep receiving updates for the Salt package, including security updates.
+
+
+# Detailed design
+[design]: #detailed-design
+
+An state to perform the migration from Salt 3000 to Salt Bundle is added to the highstate. This state checks that the 'saltversion' grain in the targeted minion and that the minion is not already running Salt Bundle. In that case, the migration is performed in the following steps:
+
+- Install the `venv-salt-minion` package.
+- Copy the minion id file, minion configuration files, PKI files and static grains for the current `salt-minion` in `/etc/salt` to the `venv-salt-minion` corresponding directories in `/etc/venv-salt-minion`.
+- Enable the `venv-salt-minion` service.
+- Disable the `salt-minion` service.
```

Comment:
What was the reason behind not removing it and just disable it? What could be the possible problems if we completely remove it?

If we don't see any major problem, ideally I would suggest removing it, just to remove the possibility that user don't go and start salt-minion service manually. As user will still be used to older way to starting/stopping minion so this could be a pretty common scenario.

Besides that, If we keep this state as part of high-state even after the migration is done, we need to make sure that there is no way we could mess a different salt-minion instance connected with a different salt master.

Reply:
Thank you Abid. This is precisely one of the points that was still subject to discussion, that's why I wrote it as one of the drawbacks. 

I was not really sure if there was any multi-master use case that would require to keep specifically the existing `salt-minion` running. In that case, `venv-salt-minion` would need a different minion id. If this is unlikely, removing it would be a better idea, as you say.

Right now I don't see the way that having this as part of the highstate could be a problem if a different master is present. However, we could only include the state as part of the highstate only if we want to trigger the migration and avoid any unexpected behaviour.

Reply:
I agree with Abid on this one. But we can put some information in the release notes if the user has multi-master configured in the salt-minion, this will be a breaking change. We should be defined what should happen in that case. But I think documentation only should be available since we shouldn't have many cases.

Reply:
At the time of the migration from salt-minion to venv-salt-minion, the customer can't have salt-minion configured to a different master. The problem that Yeray ran into was that migration+removal of salt-minion in a single step is technically not possible. Logically, it would be safe to do. 

As a result, the salt-minion needs to be removed at some time in the future, after the migration (i.e. the agent doing the removal has to be venv-salt-minion). And that's where things could be tricky, because this is when a customer could have configured salt-minion for a different Salt Master after the migration has finished. I don't expect this to happen often, but we should acknowledge this technical possibility.

All that said, the current state checks that the specified `master:` in the minion's config file is "uyuni" (looks at the pillar data for it). So, afaik, we should be safe to do the removal.

Last, but not least, let's not call this situation "multi-master". That's term is already used for a [redundant master setup in the Salt world](https://docs.saltproject.io/en/latest/topics/tutorials/multimaster.html).

Reply:
understood. Thanks for the great explanation Alex

Reply:
Thank you guys, this was helpful.

> As a result, the salt-minion needs to be removed at some time in the future.

@agraul  and when would be that time in the future?

Reply:
> Thank you guys, this was helpful.
> 
> > As a result, the salt-minion needs to be removed at some time in the future.
> 
> @agraul and when would be that time in the future?

Once the venv-salt-minion is used. The current migration start stops the salt-minion.service afaiu, the next highstate should already hit venv-salt-minion.service. I think we currently have no way to detect on the master side if a minion is using salt-minion or venv-salt-minion, so we can't use that information to trigger the removal.

Reply:
As mentioned, removing the old "salt-minion" service at the time of the migration is not really possible, as we cannot remove the old "salt" while it is actually running the migration state. This is problematic, and therefore the current manual migration state that Victor implemented is done in two steps, as our documentation already describes: one for migration and the other one for "purging" the old "salt-minion": https://documentation.suse.com/suma/4.3/en/suse-manager/client-configuration/contact-methods-saltbundle.html#_switching_from_salt_minion_to_salt_bundle

That being said, in the context of triggering this migration as part of the highstate, we are un the same situation, we cannot do this migration/removal in a single shot.

Then the question, should we trigger also the "removal" step also automatically as part of highstate, when somehow we detect that "venv-salt-minion" is running? Well.. we need to be careful here, as this could be problematic: what if a customer already configured the old "salt-minion" to use another Salt solution, we need to take care of really double checking that old "salt-minion" is not really or configured against any other Salt master before removing... I would maybe suggest that "removal" part, is NOT done automatically via highstate, but I can still be triggered manually as of today (triggering the 2nd state manually as documentation already described if customer wants)

Reply:
I get that. My only concern was around old salt-minion trying to connect again with the SUSE Manager while that machine is already being managed with the salt bundle. Would it be feasible during the migration to change or simply remove the /etc/salt/minion.de/susemanager.conf  so old minion at least doesn't point to SUSE Manager anymore unless manually done by the user?

Reply:
I guess the only option that would be a problem is if the same minion is registered to a different Uyuni/SUMA server. This is not very common, is it? In that case I'd say is safe to remove `/etc/salt/minion.d/susemanager.conf.

Reply:
Then I would vote for removing that file. Nothing further from my side.